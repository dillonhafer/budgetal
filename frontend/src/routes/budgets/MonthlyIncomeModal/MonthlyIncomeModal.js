import React, { Component } from 'react';
import PropTypes from 'prop-types';

// API
import { UpdateIncomeRequest } from '@shared/api/budgets';

// Helpers
import { error, notice } from 'window';

// Components
import { Pane, Button, Dialog } from 'evergreen-ui';
import Form, { AmountInputField } from 'components/Form';
import { Formik } from 'formik';
import * as Yup from 'yup';

const incomeValidations = Yup.object().shape({
  income: Yup.number()
    .min(1, 'Monthly Income must be at least $1.00')
    .required('Monthly Income is required'),
});

class MonthlyIncomeModal extends Component {
  static propTypes = {
    budget: PropTypes.shape({
      year: PropTypes.number.isRequired,
      month: PropTypes.number.isRequired,
      income: PropTypes.number.isRequired,
    }).isRequired,
    updateIncome: PropTypes.func.isRequired,
  };

  state = {
    visible: false,
  };

  saveIncome = async ({ income }, { setSubmitting }) => {
    try {
      const { year, month } = this.props.budget;
      const resp = await UpdateIncomeRequest({
        year,
        month,
        income,
      });
      if (resp && resp.ok) {
        this.props.updateIncome(income);
        notice('Saved Monthly Income');
      } else {
        error(resp);
      }

      this.setState({ visible: false });
    } catch (err) {
      error('Something went wrong');
    } finally {
      setSubmitting(false);
    }
  };

  renderForm = ({
    values,
    errors,
    touched,
    isSubmitting,
    handleSubmit,
    handleChange,
    handleBlur,
    handleReset,
  }) => {
    return (
      <Dialog
        preventBodyScrolling
        isShown={this.state.visible}
        title="Edit Income"
        width={400}
        onCloseComplete={() => {
          this.setState({
            visible: false,
          });
          handleReset();
        }}
        isConfirmLoading={isSubmitting}
        cancelText="Cancel"
        onConfirm={handleSubmit}
        confirmLabel="Update Income"
      >
        <Form onSubmit={handleSubmit}>
          <AmountInputField
            values={values}
            errors={errors}
            touched={touched}
            onChange={handleChange}
            onBlur={handleBlur}
            name="income"
            step={100}
            shiftStep={1000}
            label="Monthly Income"
            placeholder="(3,500.00)"
          />
        </Form>
      </Dialog>
    );
  };

  render() {
    const { budget } = this.props;
    return (
      <Pane>
        <Button
          height={32}
          appearance="primary"
          onClick={() => {
            this.setState({ visible: true });
          }}
          iconBefore="edit"
        >
          Edit Monthly Income
        </Button>

        {budget.income > 0 && (
          <Formik
            initialValues={budget}
            onSubmit={this.saveIncome}
            validationSchema={incomeValidations}
            render={this.renderForm}
          />
        )}
      </Pane>
    );
  }
}

export default MonthlyIncomeModal;
