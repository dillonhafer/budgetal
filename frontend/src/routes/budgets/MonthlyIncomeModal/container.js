import { connect } from 'react-redux';
import { updateIncome } from 'actions/budgets';
import MonthlyIncomeModal from './MonthlyIncomeModal';

export default connect(
  state => ({
    budget: state.budget.budget,
  }),
  dispatch => ({
    updateIncome: income => {
      dispatch(updateIncome(income));
    },
  }),
)(MonthlyIncomeModal);
