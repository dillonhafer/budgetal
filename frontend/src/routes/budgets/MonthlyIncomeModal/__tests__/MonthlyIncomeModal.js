import React from 'react';
import { shallow } from 'enzyme';
import MonthlyIncomeModal from '../MonthlyIncomeModal';

const defaultProps = {
  budget: {
    income: 3000.0,
    year: 2018,
    month: 8,
  },
  updateIncome: jest.fn(),
};

describe('MonthlyIncomeModal', () => {
  it('renders a MonthlyIncomeModal', () => {
    const comp = shallow(<MonthlyIncomeModal {...defaultProps} />);
    expect(comp).toMatchSnapshot();
  });
});
