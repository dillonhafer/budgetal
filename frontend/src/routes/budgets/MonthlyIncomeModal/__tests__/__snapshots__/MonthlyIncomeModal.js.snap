// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyIncomeModal renders a MonthlyIncomeModal 1`] = `
<div>
  <withTheme(Button)
    appearance="primary"
    height={32}
    iconBefore="edit"
    onClick={[Function]}
  >
    $3,000.00
  </withTheme(Button)>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "income": 3000,
        "month": 8,
        "year": 2018,
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
    validationSchema={
      ObjectSchema {
        "_blacklist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "_conditions": Array [],
        "_defaultDefault": [Function],
        "_deps": Array [],
        "_excludedEdges": Array [],
        "_exclusive": Object {},
        "_mutate": false,
        "_nodes": Array [
          "income",
        ],
        "_options": Object {
          "abortEarly": true,
          "recursive": true,
        },
        "_type": "object",
        "_typeError": [Function],
        "_whitelist": RefSet {
          "list": Set {},
          "refs": Map {},
        },
        "fields": Object {
          "income": NumberSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_deps": Array [],
            "_exclusive": Object {
              "min": true,
              "required": false,
            },
            "_mutate": false,
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "number",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "tests": Array [
              [Function],
              [Function],
            ],
            "transforms": Array [
              [Function],
            ],
          },
        },
        "tests": Array [],
        "transforms": Array [
          [Function],
        ],
      }
    }
  />
</div>
`;
